Cannot build from source anymore

It seems like we cannot build from the current master branch anymore (master/2fd53ed).

I've tested to build both on MacOS and Ubuntu20

## On MacOS

1. Try to build the provider
```
# go build -o /tmp/terraform-provider-opennebula
go: github.com/OpenNebula/one/src/oca/go/src/goca@v0.0.0-20210824074935-e55e3a53c4c0: missing go.sum entry; to add it:
	go mod download github.com/OpenNebula/one/src/oca/go/src/goca
go: github.com/OpenNebula/one/src/oca/go/src/goca@v0.0.0-20210824074935-e55e3a53c4c0: missing go.sum entry; to add it:
	go mod download github.com/OpenNebula/one/src/oca/go/src/goca
```

2. Install goca
```
# go mod download github.com/OpenNebula/one/src/oca/go/src/goca
```

3. Try to build the provider again
```
# go build -o /tmp/terraform-provider-opennebula
opennebula/helpers_vm.go:11:2: cannot find module providing package github.com/hashicorp/terraform/helper/schema: import lookup disabled by -mod=readonly
	(Go version in go.mod is 1.13, so vendor directory was not used.)
```

4. Try (fail) to install `terraform/helper/schema`
```
# go get -v github.com/hashicorp/terraform/helper/schema
go get: module github.com/hashicorp/terraform@upgrade found (v1.0.8), but does not contain package github.com/hashicorp/terraform/helper/schema
```

## On Ubuntu20

1. Try to build the provider
```
# root@server1:/tmp/git/terraform-provider-opennebula# go build -o terraform-provider-opennebula
go: downloading github.com/hashicorp/terraform-plugin-sdk v1.7.0
go: downloading github.com/OpenNebula/one/src/oca/go/src/goca v0.0.0-20210824074935-e55e3a53c4c0
<... output truncated ...>
go: finding github.com/hashicorp/terraform-config-inspect v0.0.0-20191115094559-17f92b0546e8
go: finding github.com/hashicorp/hcl v0.0.0-20170504190234-a4b07c25de5f
build github.com/OpenNebula/terraform-provider-opennebula: cannot load github.com/hashicorp/terraform/helper/schema: module github.com/hashicorp/terraform@latest found (v1.0.8), but does not contain package github.com/hashicorp/terraform/helper/schema
```

2. Try (fail) to install `terraform/helper/schema`
```
# root@server1:/tmp/git/terraform-provider-opennebula# go get -v github.com/hashicorp/terraform/helper/schema
go get github.com/hashicorp/terraform/helper/schema: module github.com/hashicorp/terraform@upgrade found (v1.0.8), but does not contain package github.com/hashicorp/terraform/helper/schema
```

Has there been some changes in the way to build the provider ?

Thanks, I can now build from source.

The new version of the plugin successfully pass `terraform init` and `terraform validate`, but does not pass `terraform plan`.

Here is the error :
```
terraform plan -compact-warnings -input=false -no-color -target=module.test1 -out=/tmp/test1.tfplan
module.test1.module.varnish1[0].random_uuid.test: Refreshing state... [id=abb43071-c724-8886-7c90-c293470e0556]
module.test1.module.vmgroup.opennebula_virtual_machine_group.vm-group-server1: Refreshing state... [id=27]
module.test1.module.securitygroup.opennebula_security_group.internal: Refreshing state... [id=123]
module.test1.module.securitygroup.opennebula_security_group.server1: Refreshing state... [id=124]

Warning: Resource targeting is in effect

You are creating a plan with the -target option, which means that the result of this plan may not represent all of the changes requested by the current configuration.

The -target option is not for routine use, and is provided only for exceptional situations such as recovering from errors or mistakes, or when Terraform specifically suggests to use it as part of an error message.

Error: Plugin did not respond

  with module.test1.module.securitygroup.opennebula_security_group.server1,
  on modules/securitygroup/main.tf line 40, in resource "opennebula_security_group" "server1":
  40: resource "opennebula_security_group" "server1" {

The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ReadResource call. The plugin logs may contain more details.

Stack trace from the terraform-provider-opennebula plugin:

panic: interface conversion: interface {} is int, not string

goroutine 253 [running]:
github.com/OpenNebula/terraform-provider-opennebula/opennebula.matchNIC(0xc000559ecc, {{{{0x0, 0x0}, {0xc0004ae22b, 0x3}}, {0xc000242c00, 0xf, 0x10}}})
	/Users/thomas/github/OpenNebula/terraform-provider-opennebula/opennebula/resource_opennebula_virtual_machine.go:765 +0x694
github.com/OpenNebula/terraform-provider-opennebula/opennebula.flattenVMNIC(0xc000306230, 0x1928d88)
	/Users/thomas/github/OpenNebula/terraform-provider-opennebula/opennebula/resource_opennebula_virtual_machine.go:811 +0x512
github.com/OpenNebula/terraform-provider-opennebula/opennebula.resourceOpennebulaVirtualMachineReadCustom(0x1901f40, {0x19182e0, 0xc0002e33e0}, 0x196d640, 0x196d648)
	/Users/thomas/github/OpenNebula/terraform-provider-opennebula/opennebula/resource_opennebula_virtual_machine.go:510 +0x3b6
github.com/OpenNebula/terraform-provider-opennebula/opennebula.resourceOpennebulaVirtualMachineRead(...)
	/Users/thomas/github/OpenNebula/terraform-provider-opennebula/opennebula/resource_opennebula_virtual_machine.go:533
github.com/OpenNebula/terraform-provider-opennebula/opennebula.resourceOpennebulaVirtualMachineExists(0xc000306230, {0x19182e0, 0xc0002e33e0})
	/Users/thomas/github/OpenNebula/terraform-provider-opennebula/opennebula/resource_opennebula_virtual_machine.go:875 +0x3a
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc00052e600, 0xc0005140a0, {0x19182e0, 0xc0002e33e0})
	/Users/thomas/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.7.0/helper/schema/resource.go:440 +0x131
github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ReadResource(0xc000010ae8, {0xc0004eed70, 0x108ebc6}, 0xc0004eed70)
	/Users/thomas/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.7.0/internal/helper/plugin/grpc_provider.go:525 +0x34b
github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ReadResource_Handler({0x18f82a0, 0xc000010ae8}, {0x1a9e130, 0xc00064e210}, 0xc00068f260, 0x0)
	/Users/thomas/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.7.0/internal/tfplugin5/tfplugin5.pb.go:3153 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0000d14a0, {0x1aab008, 0xc000188180}, 0xc0006a4200, 0xc00054cff0, 0x2005b70, 0x0)
	/Users/thomas/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:995 +0xe0d
google.golang.org/grpc.(*Server).handleStream(0xc0000d14a0, {0x1aab008, 0xc000188180}, 0xc0006a4200, 0x0)
	/Users/thomas/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:1275 +0xa56
google.golang.org/grpc.(*Server).serveStreams.func1.1()
	/Users/thomas/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:710 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/Users/thomas/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:708 +0xef

Error: The terraform-provider-opennebula plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

One thing that looks strange is that terraform is showing the path of where I have cloned the github repo (`/Users/thomas/github/OpenNebula/terraform-provider-opennebula`), but I'm not running terraform commands in this folder.

--- a/opennebula/helpers.go
+++ b/opennebula/helpers.go
@@ -66,6 +66,10 @@ func LockLevelToString(lock int) string {
 	return ""
 }
 
+func emptyOrEqual(config interface{}, value interface{}) bool {
+	return isEmptyValue(reflect.ValueOf(config)) || value == config
+}
+
 func isEmptyValue(v reflect.Value) bool {
 	switch v.Kind() {
 	case reflect.String, reflect.Array, reflect.Map, reflect.Slice:

--- a/opennebula/resource_opennebula_virtual_machine.go
+++ b/opennebula/resource_opennebula_virtual_machine.go
@@ -617,6 +617,8 @@ func flattenVMDisk(d *schema.ResourceData, vmTemplate *vm.Template) error {
 
 	// Set disks to resource
 	disks := vmTemplate.GetDisks()
+	diskConfigs := d.Get("disk").([]interface{})
+
 	diskList := make([]interface{}, 0, len(disks))
 
 diskLoop:
@@ -633,15 +635,11 @@ diskLoop:
 		}
 
 		// copy disk config values
-		var match bool
 		var diskMap map[string]interface{}
 
-		diskConfigs := d.Get("disk").([]interface{})
-
+		match := false
 		for j := 0; j < len(diskConfigs); j++ {
 			diskConfig := diskConfigs[j].(map[string]interface{})
-
-			match = false
 
 			// try to reidentify the disk based on it's configuration values
 			if !matchDisk(diskConfig, disk) {
@@ -1021,6 +1019,8 @@ func resourceOpennebulaVirtualMachineUpdate(d *schema.ResourceData, meta interfa
 		timeout := d.Get("timeout").(int)
 
 		// get unique elements of each list of configs
+		// NOTE: diffListConfig relies on Set, so we may loose list ordering of disks here
+		// it's why we reorder the attach list below
 		toDetach, toAttach := diffListConfig(newDisksCfg, attachedDisksCfg,
 			&schema.Resource{
 				Schema: diskFields(),

--- a/opennebula/resource_opennebula_virtual_machine.go
+++ b/opennebula/resource_opennebula_virtual_machine.go
@@ -711,7 +711,7 @@ func flattenVMNICComputed(NICConfig map[string]interface{}, NIC shared.NIC) map[
 	if len(NICConfig["physical_device"].(string)) > 0 {
 		NICMap["physical_device"] = NICMap["computed_physical_device"]
 	}
-	if len(NICConfig["security_groups"].(string)) > 0 {
+	if len(NICConfig["security_groups"].([]interface{})) > 0 {
 		NICMap["security_groups"] = NICMap["computed_security_groups"]
 	}
 

--- a/opennebula/resource_opennebula_virtual_machine.go
+++ b/opennebula/resource_opennebula_virtual_machine.go
@@ -1004,7 +1004,7 @@ func resourceOpennebulaVirtualMachineUpdate(d *schema.ResourceData, meta interfa
 		_, err = waitForVMState(vmc, timeout, "RUNNING")
 		if err != nil {
 			return fmt.Errorf(
-				"waiting for virtual machine (ID:%d) to be in state %s: %s", vmc.ID, strings.Join(vmDiskUpdateReadyStates, " "), err)
+				"waiting for virtual machine (ID:%d) to be in state %s: %s", vmc.ID, "RUNNING", err)
 		}
 	}
 
@@ -1307,7 +1307,7 @@ func resourceOpennebulaVirtualMachineUpdate(d *schema.ResourceData, meta interfa
 		_, err = waitForVMState(vmc, timeout, "RUNNING")
 		if err != nil {
 			return fmt.Errorf(
-				"waiting for virtual machine (ID:%d) to be in state %s: %s", vmc.ID, strings.Join(vmDiskUpdateReadyStates, " "), err)
+				"waiting for virtual machine (ID:%d) to be in state %s: %s", vmc.ID, "RUNNING", err)
 		}
 
 		log.Printf("[INFO] Update VM configuration: %s", tpl.String())
@@ -1320,7 +1320,7 @@ func resourceOpennebulaVirtualMachineUpdate(d *schema.ResourceData, meta interfa
 		_, err = waitForVMState(vmc, timeout, "RUNNING")
 		if err != nil {
 			return fmt.Errorf(
-				"waiting for virtual machine (ID:%d) to be in state %s: %s", vmc.ID, strings.Join(vmDiskUpdateReadyStates, " "), err)
+				"waiting for virtual machine (ID:%d) to be in state %s: %s", vmc.ID, "RUNNING", err)
 		}
 	}
 
@@ -1360,7 +1360,7 @@ func resourceOpennebulaVirtualMachineDelete(d *schema.ResourceData, meta interfa
 	_, err = waitForVMState(vmc, timeout, "RUNNING")
 	if err != nil {
 		return fmt.Errorf(
-			"waiting for virtual machine (ID:%d) to be in state %s: %s", vmc.ID, strings.Join(vmDiskUpdateReadyStates, " "), err)
+			"waiting for virtual machine (ID:%d) to be in state %s: %s", vmc.ID, "RUNNING", err)
 	}
 
 	if err = vmc.TerminateHard(); err != nil {

--- a/opennebula/resource_opennebula_virtual_machine.go
+++ b/opennebula/resource_opennebula_virtual_machine.go
@@ -1158,40 +1158,41 @@ func resourceOpennebulaVirtualMachineUpdate(d *schema.ResourceData, meta interfa
 				}
 			}
 
-			// reorder toAttach NIC list according to new list order
-			newNICtoAttach := make([]interface{}, len(toAttach))
-			i := 0
-			for _, newNICIf := range newNicsCfg {
-				newNIC := newNICIf.(map[string]interface{})
-				newNICSecGroup := newNIC["security_groups"].([]interface{})
+		}
 
-				for _, NICToAttachIf := range toAttach {
-					NIC := NICToAttachIf.(map[string]interface{})
+		// reorder toAttach NIC list according to new nics list order
+		newNICtoAttach := make([]interface{}, len(toAttach))
+		i := 0
+		for _, newNICIf := range newNicsCfg {
+			newNIC := newNICIf.(map[string]interface{})
+			newNICSecGroup := newNIC["security_groups"].([]interface{})
 
-					// if NIC have the same attributes
+			for _, NICToAttachIf := range toAttach {
+				NIC := NICToAttachIf.(map[string]interface{})
 
-					// compare security_groups
-					NICSecGroup := NIC["security_groups"].([]interface{})
+				// if NIC have the same attributes
 
-					if ArrayToString(NICSecGroup, ",") != ArrayToString(newNICSecGroup, ",") {
-						continue
-					}
+				// compare security_groups
+				NICSecGroup := NIC["security_groups"].([]interface{})
 
-					// compare other attributes
-					if (NIC["ip"] == newNIC["ip"] &&
-						NIC["mac"] == newNIC["mac"]) &&
-						NIC["model"] == newNIC["model"] &&
-						NIC["virtio_queues"] == newNIC["virtio_queues"] &&
-						NIC["physical_device"] == newNIC["physical_device"] {
+				if ArrayToString(NICSecGroup, ",") != ArrayToString(newNICSecGroup, ",") {
+					continue
+				}
 
-						newNICtoAttach[i] = NIC
-						i++
-						break
-					}
+				// compare other attributes
+				if (NIC["ip"] == newNIC["ip"] &&
+					NIC["mac"] == newNIC["mac"]) &&
+					NIC["model"] == newNIC["model"] &&
+					NIC["virtio_queues"] == newNIC["virtio_queues"] &&
+					NIC["physical_device"] == newNIC["physical_device"] {
+
+					newNICtoAttach[i] = NIC
+					i++
+					break
 				}
 			}
-			toAttach = newNICtoAttach
 		}
+		toAttach = newNICtoAttach
 
 		// Detach the nics
 		for _, nicIf := range toDetach {

--- a/opennebula/resource_opennebula_virtual_machine.go
+++ b/opennebula/resource_opennebula_virtual_machine.go
@@ -1029,6 +1029,28 @@ func resourceOpennebulaVirtualMachineUpdate(d *schema.ResourceData, meta interfa
 			"target",
 			"driver")
 
+		// reorder toAttach disk list according to new disks list order
+		newDisktoAttach := make([]interface{}, len(toAttach))
+		i := 0
+		for _, newDiskIf := range newDisksCfg {
+			newDisk := newDiskIf.(map[string]interface{})
+
+			for _, diskToAttachIf := range toAttach {
+				disk := diskToAttachIf.(map[string]interface{})
+
+				// if disk have the same attributes
+				if (disk["target"] == newDisk["target"]) &&
+					disk["size"] == newDisk["size"] &&
+					disk["driver"] == newDisk["driver"] {
+
+					newDisktoAttach[i] = disk
+					i++
+					break
+				}
+			}
+		}
+		toAttach = newDisktoAttach
+
 		// get disks to resize
 		_, toResize := diffListConfig(newDisksCfg, attachedDisksCfg,
 			&schema.Resource{

--- a/opennebula/resource_opennebula_virtual_machine.go
+++ b/opennebula/resource_opennebula_virtual_machine.go
@@ -1057,6 +1057,7 @@ func resourceOpennebulaVirtualMachineUpdate(d *schema.ResourceData, meta interfa
 			&schema.Resource{
 				Schema: diskFields(),
 			},
+			"image_id",
 			"size")
 
 		// Detach the disks

--- a/opennebula/resource_opennebula_virtual_machine.go
+++ b/opennebula/resource_opennebula_virtual_machine.go
@@ -1198,7 +1198,7 @@ func resourceOpennebulaVirtualMachineUpdate(d *schema.ResourceData, meta interfa
 		for _, nicIf := range toDetach {
 			nicConfig := nicIf.(map[string]interface{})
 
-			nicID := nicConfig["network_id"].(int)
+			nicID := nicConfig["nic_id"].(int)
 
 			err := vmNICDetach(vmc, timeout, nicID)
 			if err != nil {

--- a/opennebula/resource_opennebula_virtual_machine.go
+++ b/opennebula/resource_opennebula_virtual_machine.go
@@ -593,9 +593,21 @@ func matchDisk(diskConfig map[string]interface{}, disk shared.Disk) bool {
 	driver, _ := disk.Get(shared.Driver)
 	target, _ := disk.Get(shared.TargetDisk)
 
-	return (len(diskConfig["target"].(string)) == 0 || target != diskConfig["computed_target"].(string)) ||
-		(diskConfig["size"].(int) == 0 || size != diskConfig["computed_size"].(int)) ||
-		(len(diskConfig["driver"].(string)) == 0 || driver != diskConfig["computed_driver"].(string))
+	return emptyOrEqual(diskConfig["target"], target) &&
+		emptyOrEqual(diskConfig["size"], size) &&
+		emptyOrEqual(diskConfig["driver"], driver)
+
+}
+
+func matchDiskComputed(diskConfig map[string]interface{}, disk shared.Disk) bool {
+
+	size, _ := disk.GetI(shared.Size)
+	driver, _ := disk.Get(shared.Driver)
+	target, _ := disk.Get(shared.TargetDisk)
+
+	return (target == diskConfig["computed_target"].(string)) &&
+		(size == diskConfig["computed_size"].(int)) &&
+		(driver == diskConfig["computed_driver"].(string))
 
 }
 
@@ -615,7 +627,7 @@ diskLoop:
 		for _, tplDiskConfigIf := range tplDiskConfigs {
 			tplDiskConfig := tplDiskConfigIf.(map[string]interface{})
 
-			if matchDisk(tplDiskConfig, disk) {
+			if matchDiskComputed(tplDiskConfig, disk) {
 				continue diskLoop
 			}
 		}

--- a/opennebula/resource_opennebula_virtual_machine.go
+++ b/opennebula/resource_opennebula_virtual_machine.go
@@ -705,6 +705,9 @@ func flattenVMNICComputed(NICConfig map[string]interface{}, NIC shared.NIC) map[
 	if len(NICConfig["mac"].(string)) > 0 {
 		NICMap["mac"] = NICMap["computed_mac"]
 	}
+	if len(NICConfig["model"].(string)) > 0 {
+		NICMap["model"] = NICMap["computed_model"]
+	}
 	if len(NICConfig["virtio_queues"].(string)) > 0 {
 		NICMap["virtio_queues"] = NICMap["computed_virtio_queues"]
 	}

--- a/opennebula/resource_opennebula_virtual_machine.go
+++ b/opennebula/resource_opennebula_virtual_machine.go
@@ -755,24 +755,65 @@ func matchNIC(NICConfig map[string]interface{}, NIC shared.NIC) bool {
 	virtioQueues, _ := NIC.GetStr("VIRTIO_QUEUES")
 	securityGroupsArray, _ := NIC.Get(shared.SecurityGroups)
 
-	sgStr := strings.Split(securityGroupsArray, ",")
+	if NICConfig["security_groups"] != nil && len(NICConfig["security_groups"].([]interface{})) > 0 {
+
+		sg := strings.Split(securityGroupsArray, ",")
+		sgConfig := NICConfig["security_groups"].([]interface{})
+
+		if len(sg) != len(sgConfig) {
+			return false
+		}
+
+		for i := 0; i < len(sg); i++ {
+			sgInt, err := strconv.ParseInt(sg[i], 10, 0)
+			if err != nil {
+				return false
+			}
+			if int(sgInt) != sgConfig[i].(int) {
+				return false
+			}
+		}
+
+	}
+
+	return emptyOrEqual(NICConfig["ip"], ip) &&
+		emptyOrEqual(NICConfig["mac"], mac) &&
+		emptyOrEqual(NICConfig["physical_device"], physicalDevice) &&
+		emptyOrEqual(NICConfig["model"], model) &&
+		emptyOrEqual(NICConfig["virtio_queues"], virtioQueues)
+}
+
+func matchNICComputed(NICConfig map[string]interface{}, NIC shared.NIC) bool {
+	ip, _ := NIC.Get(shared.IP)
+	mac, _ := NIC.Get(shared.MAC)
+	physicalDevice, _ := NIC.GetStr("PHYDEV")
+
+	model, _ := NIC.Get(shared.Model)
+	virtioQueues, _ := NIC.GetStr("VIRTIO_QUEUES")
+	securityGroupsArray, _ := NIC.Get(shared.SecurityGroups)
+
+	sg := strings.Split(securityGroupsArray, ",")
 	sgConfig := NICConfig["computed_security_groups"].([]interface{})
-	if len(sgStr) != len(sgConfig) {
+
+	if len(sg) != len(sgConfig) {
 		return false
 	}
 
-	for i := 0; i < len(sgStr); i++ {
-		if sgStr[i] != sgConfig[i].(string) {
+	for i := 0; i < len(sg); i++ {
+		sgInt, err := strconv.ParseInt(sg[i], 10, 0)
+		if err != nil {
+			return false
+		}
+		if int(sgInt) != sgConfig[i].(int) {
 			return false
 		}
 	}
 
-	return len(NICConfig["ip"].(string)) == 0 || ip == NICConfig["computed_ip"].(string) &&
-		len(NICConfig["mac"].(string)) == 0 && mac == NICConfig["computed_mac"].(string) &&
-		len(NICConfig["physical_device"].(string)) == 0 && physicalDevice == NICConfig["computed_physical_device"].(string) &&
-		len(NICConfig["model"].(string)) == 0 && model == NICConfig["computed_model"].(string) &&
-		len(NICConfig["virtio_queues"].(string)) == 0 && virtioQueues == NICConfig["computed_virtio_queues"].(string)
-
+	return ip == NICConfig["computed_ip"].(string) &&
+		mac == NICConfig["computed_mac"].(string) &&
+		physicalDevice == NICConfig["computed_physical_device"].(string) &&
+		model == NICConfig["computed_model"].(string) &&
+		virtioQueues == NICConfig["computed_virtio_queues"].(string)
 }
 
 // flattenVMNIC is similar to flattenNIC but deal with computed_* attributes
@@ -791,7 +832,7 @@ NICLoop:
 		for _, tplNICConfigIf := range tplNICConfigs {
 			tplNICConfig := tplNICConfigIf.(map[string]interface{})
 
-			if matchNIC(tplNICConfig, nic) {
+			if matchNICComputed(tplNICConfig, nic) {
 				continue NICLoop
 			}
 		}

--- a/opennebula/resource_opennebula_virtual_machine.go
+++ b/opennebula/resource_opennebula_virtual_machine.go
@@ -21,6 +21,7 @@ var (
 	vmDiskUpdateReadyStates = []string{"RUNNING", "POWEROFF"}
 	vmDiskResizeReadyStates = []string{"RUNNING", "POWEROFF", "UNDEPLOYED"}
 	vmNICUpdateReadyStates  = vmDiskUpdateReadyStates
+	vmDeleteReadyStates     = []string{"RUNNING", "HOLD"}
 )
 
 type flattenVMPart func(d *schema.ResourceData, vmTemplate *vm.Template) error
@@ -1357,10 +1358,10 @@ func resourceOpennebulaVirtualMachineDelete(d *schema.ResourceData, meta interfa
 	// wait state to be ready
 	timeout := d.Get("timeout").(int)
 
-	_, err = waitForVMState(vmc, timeout, "RUNNING")
+	_, err = waitForVMState(vmc, timeout, vmDeleteReadyStates...)
 	if err != nil {
 		return fmt.Errorf(
-			"waiting for virtual machine (ID:%d) to be in state %s: %s", vmc.ID, "RUNNING", err)
+			"waiting for virtual machine (ID:%d) to be in state %s: %s", vmc.ID, strings.Join(vmDeleteReadyStates, " "), err)
 	}
 
 	if err = vmc.TerminateHard(); err != nil {

--- a/opennebula/resource_opennebula_virtual_machine.go
+++ b/opennebula/resource_opennebula_virtual_machine.go
@@ -822,6 +822,8 @@ func flattenVMNIC(d *schema.ResourceData, vmTemplate *vm.Template) error {
 
 	// Set Nics to resource
 	nics := vmTemplate.GetNICs()
+	nicsConfigs := d.Get("nic").([]interface{})
+
 	nicList := make([]interface{}, 0, len(nics))
 
 NICLoop:
@@ -838,14 +840,11 @@ NICLoop:
 		}
 
 		// copy nic config values
-		var match bool
 		var nicMap map[string]interface{}
 
-		nicsConfigs := d.Get("nic").([]interface{})
+		match := false
 		for j := 0; j < len(nicsConfigs); j++ {
 			nicConfig := nicsConfigs[j].(map[string]interface{})
-
-			match = false
 
 			// try to reidentify the nic based on it's configuration values
 			// network_id is not sufficient in case of a network attached twice
@@ -1100,6 +1099,8 @@ func resourceOpennebulaVirtualMachineUpdate(d *schema.ResourceData, meta interfa
 		timeout := d.Get("timeout").(int)
 
 		// get unique elements of each list of configs
+		// NOTE: diffListConfig relies on Set, so we may loose list ordering of NICs here
+		// it's why we reorder the attach list below
 		toDetach, toAttach := diffListConfig(newNicsCfg, attachedNicsCfg,
 			&schema.Resource{
 				Schema: nicFields(),
@@ -1112,6 +1113,9 @@ func resourceOpennebulaVirtualMachineUpdate(d *schema.ResourceData, meta interfa
 			"virtio_queues",
 			"physical_device")
 
+		// in case of NICs updated in the middle of the NIC list
+		// they would be reattached at the end of the list (we don't have in place XML-RPC update method).
+		// keep_nic_order prevent this behavior adding more NICs to detach/attach to keep initial ordering
 		if d.Get("keep_nic_order").(bool) && len(toDetach) > 0 {
 
 			// retrieve the minimal nic ID to detach

--- a/opennebula/resource_opennebula_virtual_machine_test.go
+++ b/opennebula/resource_opennebula_virtual_machine_test.go
@@ -264,48 +264,48 @@ func TestAccVirtualMachineNICUpdate(t *testing.T) {
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "name", "test-virtual_machine"),
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.#", "2"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.0.computed_ip", "172.16.100.111"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.1.computed_ip", "172.16.100.131"),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.0.computed_ip", "172.16.100.131"),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.1.computed_ip", "172.16.100.111"),
 				),
 			},
 			{
 				Config: testAccVirtualMachineTemplateConfigNICIPUpdate,
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "name", "test-virtual_machine"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.#", "1"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.0.computed_ip", "172.16.100.112"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.1.computed_ip", "172.16.100.131"),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.#", "2"),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.0.computed_ip", "172.16.100.131"),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.1.computed_ip", "172.16.100.112"),
 				),
 			},
 			{
 				Config: testAccVirtualMachineTemplateConfigMultipleNICs,
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "name", "test-virtual_machine"),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "keep_nic_order", "false"),
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.#", "4"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.keep_nic_order", "false"),
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.0.computed_ip", "172.16.100.112"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.0.computed_ip", "172.16.100.132"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.0.computed_ip", "172.16.100.113"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.0.computed_ip", "172.16.100.133"),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.1.computed_ip", "172.16.100.132"),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.2.computed_ip", "172.16.100.113"),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.3.computed_ip", "172.16.100.133"),
 				),
 			},
 			{
 				Config: testAccVirtualMachineTemplateConfigMultipleNICsOrderedUpdate,
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "name", "test-virtual_machine"),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "keep_nic_order", "true"),
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.#", "4"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.keep_nic_order", "true"),
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.0.computed_ip", "172.16.100.112"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.0.computed_ip", "172.16.100.134"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.0.computed_ip", "172.16.100.113"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.0.computed_ip", "172.16.100.133"),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.1.computed_ip", "172.16.100.134"),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.2.computed_ip", "172.16.100.113"),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.3.computed_ip", "172.16.100.133"),
 				),
 			},
 			{
 				Config: testAccVirtualMachineTemplateConfigNICDetached,
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "name", "test-virtual_machine"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.keep_nic_order", "false"),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "keep_nic_order", "false"),
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "nic.#", "0"),
 				),
 			},
@@ -1149,8 +1149,8 @@ var testAccVirtualMachineTemplateConfigNICUpdate = testNICVNetResources + `
 		ip = "172.16.100.131"
 	  }
 	  nic {
-		  network_id = opennebula_virtual_network.network2.id
-		  ip = "172.16.100.111"
+		network_id = opennebula_virtual_network.network2.id
+		ip = "172.16.100.111"
 	  }
 	
 	  timeout = 5
@@ -1192,10 +1192,11 @@ var testAccVirtualMachineTemplateConfigNICIPUpdate = testNICVNetResources + `
 		ip = "172.16.100.131"
 	  }
 	  nic {
-		  network_id = opennebula_virtual_network.network2.id
-		  ip = "172.16.100.112"
+		network_id = opennebula_virtual_network.network2.id
+		ip = "172.16.100.112"
 	  }
-	
+
+
 	  timeout = 5
 }
 `
@@ -1230,6 +1231,7 @@ var testAccVirtualMachineTemplateConfigMultipleNICs = testNICVNetResources + `
 		customer = "test"
 	  }
 
+	  keep_nic_order = false
 	  nic {
 		network_id = opennebula_virtual_network.network2.id
 		ip         = "172.16.100.112"
@@ -1332,6 +1334,8 @@ resource "opennebula_virtual_machine" "test" {
 	  env = "prod"
 	  customer = "test"
 	}
+
+	keep_nic_order = false
 
 	timeout = 5
 }

--- a/opennebula/resource_opennebula_virtual_network.go
+++ b/opennebula/resource_opennebula_virtual_network.go
@@ -717,12 +717,11 @@ func matchARs(ARConfig map[string]interface{}, AR vn.AR) bool {
 
 	return AR.Type == ARConfig["ar_type"].(string) &&
 		AR.Size == ARConfig["size"].(int) &&
-		len(ARConfig["ip4"].(string)) > 0 || AR.IP == ARConfig["ip4"].(string) &&
-		len(ARConfig["ip6"].(string)) > 0 || AR.IP6 == ARConfig["ip6"].(string) &&
-		len(ARConfig["mac"].(string)) > 0 || AR.MAC == ARConfig["mac"].(string) &&
-		len(ARConfig["global_prefix"].(string)) > 0 || AR.GlobalPrefix == ARConfig["global_prefix"].(string) &&
-		len(ARConfig["ula_prefix"].(string)) > 0 || AR.ULAPrefix == ARConfig["ula_prefix"].(string)
-
+		emptyOrEqual(ARConfig["ip4"], AR.IP) &&
+		emptyOrEqual(ARConfig["ip6"], AR.IP6) &&
+		emptyOrEqual(ARConfig["mac"], AR.MAC) &&
+		emptyOrEqual(ARConfig["global_prefix"], AR.GlobalPrefix) &&
+		emptyOrEqual(ARConfig["ula_prefix"], AR.ULAPrefix)
 }
 
 func resourceOpennebulaVirtualNetworkRead(d *schema.ResourceData, meta interface{}) error {

--- a/opennebula/resource_opennebula_virtual_network.go
+++ b/opennebula/resource_opennebula_virtual_network.go
@@ -773,10 +773,25 @@ func resourceOpennebulaVirtualNetworkRead(d *schema.ResourceData, meta interface
 		return err
 	}
 
+	err = flattenVnetARs(d, vn)
+	if err != nil {
+		return err
+	}
+
+	if vn.Lock != nil {
+		d.Set("lock", LockLevelToString(vn.Lock.Locked))
+	}
+
+	return nil
+}
+
+func flattenVnetARs(d *schema.ResourceData, vn *vn.VirtualNetwork) error {
+
 	ARSet := make([]map[string]interface{}, 0, len(vn.ARs))
+	ARConfigs := d.Get("ar").(*schema.Set).List()
+
 	for _, AR := range vn.ARs {
 
-		ARConfigs := d.Get("ar").(*schema.Set).List()
 		match := false
 
 		// retrieve the associated AR config
@@ -803,10 +818,6 @@ func resourceOpennebulaVirtualNetworkRead(d *schema.ResourceData, meta interface
 
 	if err := d.Set("ar", ARSet); err != nil {
 		log.Printf("[WARN] Error setting ar for Virtual Network %x, error: %s", vn.ID, err)
-	}
-
-	if vn.Lock != nil {
-		d.Set("lock", LockLevelToString(vn.Lock.Locked))
 	}
 
 	return nil

--- a/opennebula/shared_schemas.go
+++ b/opennebula/shared_schemas.go
@@ -263,6 +263,11 @@ func makeDiskVector(diskConfig map[string]interface{}) *shared.Disk {
 
 	for k, v := range diskConfig {
 
+		if k == "image_id" {
+			disk.Add(shared.ImageID, strconv.Itoa(v.(int)))
+			continue
+		}
+
 		if isEmptyValue(reflect.ValueOf(v)) {
 			continue
 		}
@@ -274,8 +279,6 @@ func makeDiskVector(diskConfig map[string]interface{}) *shared.Disk {
 			disk.Add(shared.Driver, v.(string))
 		case "size":
 			disk.Add(shared.Size, strconv.Itoa(v.(int)))
-		case "image_id":
-			disk.Add(shared.ImageID, strconv.Itoa(v.(int)))
 		}
 	}
 

